source/opt/folding_rules.cpp

```diff
@@ -124,6 +124,102 @@ Instruction* NonConstInput(IRContext* context, const analysis::Constant* c,
+      words = result.GetWords();
+    } else {
+      utils::FloatProxy<float> result(float_constant->GetFloat());
+      words = result.GetWords();
+    }
+  } else if (const auto* int_constant = c->AsIntConstant()) {
+    uint32_t width = int_constant->type()->AsInteger()->width();
+    assert(width == 32 || width == 64);
+    if (width == 64) {
+      uint64_t uval = static_cast<uint64_t>(int_constant->GetU64());
+      words = ExtractInts(uval);
+    } else {
+      words.push_back(static_cast<uint32_t>(int_constant->GetU32()));
+    }
+  } else if (const auto* vec_constant = c->AsVectorConstant()) {
+    if (const auto* vec_type = vec_constant->type()->AsVector()) {
+      if (vec_type->element_type()->AsInteger() == nullptr &&
+          vec_type->element_type()->AsFloat() == nullptr) {
+        return words;
+      }
+    } else {
+      return words;
+    }
+
+    for (const auto* comp : vec_constant->GetComponents()) {
+      auto comp_in_words =
+          GetWordsFromNumericScalarOrVectorConstant(const_mgr, comp);
+      words.insert(words.end(), comp_in_words.begin(), comp_in_words.end());
+    }
+  }
+  return words;
+}
+
+const analysis::Constant* ConvertWordsToNumericScalarOrVectorConstant(
+    analysis::ConstantManager* const_mgr, const std::vector<uint32_t>& words,
+    const analysis::Type* type) {
+  if (type->AsInteger() || type->AsFloat())
+    return const_mgr->GetConstant(type, words);
+
+  if (const auto* vec_type = type->AsVector()) {
+    const auto* element_type = vec_type->element_type();
+    uint32_t width_in_bits = 0;
+    if (const auto* float_type = element_type->AsFloat())
+      width_in_bits = float_type->width();
+    else if (const auto* int_type = element_type->AsInteger())
+      width_in_bits = int_type->width();
+
+    if (width_in_bits == 32) {
+      if (vec_type->element_count() != static_cast<uint32_t>(words.size()))
+        return nullptr;
+
+      std::vector<uint32_t> element_ids;
+      for (uint32_t i = 0; i < vec_type->element_count(); ++i) {
+        const analysis::Constant* element_constant =
+            const_mgr->GetConstant(element_type, {words[i]});
+        auto element_id =
+            const_mgr->GetDefiningInstruction(element_constant)->result_id();
+        element_ids.push_back(element_id);
+      }
+      return const_mgr->GetConstant(type, element_ids);
+    }
+
+    if (width_in_bits == 64) {
+      if (2 * vec_type->element_count() != static_cast<uint32_t>(words.size()))
+        return nullptr;
+
+      std::vector<uint32_t> element_ids;
+      for (uint32_t i = 0; i < vec_type->element_count(); ++i) {
+        const analysis::Constant* element_constant = const_mgr->GetConstant(
+            element_type, {words[2 * i], words[2 * i + 1]});
+        auto element_id =
+            const_mgr->GetDefiningInstruction(element_constant)->result_id();
+        element_ids.push_back(element_id);
+      }
+      return const_mgr->GetConstant(type, element_ids);
+    }
```

Comment:
This code is almost the same as the 32 bit case.  You could probably write it more generically, and avoid the repetition.  For example, you could have a variable called `words_per_element` which could be either 0 or 1.  You can extract the sub vector from `words` using 

```
auto first_word = words.begin()+(words_per_element*i);
std::vector<uint32_t> const_data(first_word, first_word+words_per_element);
```

Reply:
done